### PR TITLE
update: forcing explicit optional types

### PR DESCRIPTION
**Pull request by krneta**
*Issue #, if available:*

*Description of changes:*

*Testing done:*

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [ ] I have read the [CONTRIBUTING](https://github.com/aws/amazon-braket-containers/blob/main/CONTRIBUTING.md) doc
- [ ] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/amazon-braket-containers/blob/main/CONTRIBUTING.md#commit-your-change)
- [ ] I have updated any necessary documentation, including [READMEs](https://github.com/aws/amazon-braket-containers/blob/main/README.md) and [API docs](https://github.com/aws/amazon-braket-containers/blob/main/CONTRIBUTING.md#documentation-guidelines) (if appropriate)

#### Tests

- [ ] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [ ] I have checked that my tests are not configured for a specific region or account (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/braket/flake8_plugins/braket_checkstyle_plugin.py [hunk 1]
```diff
@@ -87,6 +87,7 @@ class _Visitor(ast.NodeVisitor):
         "BCS020": "Function '%s' has return documentation but no return type.",
         "BCS021": "Function '%s' is missing return documentation.",
         "BCS022": "Found '%d' invalid indents starting with line ('%s').",
+        "BCS023": "Argument '%s' is optional but type hint doesn't end with '| None'.",
```
Comment (ajberdy) [inline src/braket/flake8_plugins/braket_checkstyle_plugin.py:90]: I'd prefer we avoid "optional" here as it is ambiguous and doesn't exactly align with the requirements. `bool = False` or `str = ''` are both optional args that don't include `None` in their type. It's an existing pain point that `Optional[...]` doesn't align with whether an arg is "optional", so might as well use clearer terms

File: src/braket/flake8_plugins/braket_checkstyle_plugin.py [hunk 2]
```diff
@@ -278,28 +279,46 @@ def _check_annotation(
...
                     node=node,
                     code="BCS005",
                     arguments=(arg_name, annotation_doc, documented_type),
                 )
+        if (
+            default_value
+            and isinstance(default_value, ast.Constant)
+            and default_value.value is None
+        ):
+            if not documented_type.endswith("|None") and not documented_type.startswith("Optional"):
```
Comment (ajberdy) [inline src/braket/flake8_plugins/braket_checkstyle_plugin.py:307]: Is there a way to do this through attributes without string processing? Seems this allows `str | None` and `Optional[str]`, but what about `Union[str, None]` or `None | str`?

File: test/unit_tests/braket/flake8_plugins/example_files/missing_doc.py [hunk 3]
```diff
@@ -1,3 +1,16 @@
...
+    """This is a description.
+    Args:
+        a0 (int): This is a parameter
+        a1 (str|None): This is an optional parameter
+        a2 (str): This is an optional parameter missing optional type hint
+        a3 (str|None): This is an optional parameter
+    Returns:
+        int: value of the function
+    """
+    pass
```
Comment (ajberdy) [inline test/unit_tests/braket/flake8_plugins/example_files/missing_doc.py:16]: nit: newline at eof

File: src/braket/flake8_plugins/braket_checkstyle_plugin.py (hunk 1 above)
Comment (krneta) [reply src/braket/flake8_plugins/braket_checkstyle_plugin.py:90]: Good call. Will change.

File: src/braket/flake8_plugins/braket_checkstyle_plugin.py (hunk 2 above)
Comment (krneta) [reply src/braket/flake8_plugins/braket_checkstyle_plugin.py:307]: I think it makes sense to standardize on putting | None last, and we'll see if we need to deal with edge cases for Union.

File: test/unit_tests/braket/flake8_plugins/example_files/missing_doc.py (hunk 3 above)
Comment (krneta) [reply test/unit_tests/braket/flake8_plugins/example_files/missing_doc.py:16]: Will do, thanks!

File: src/braket/flake8_plugins/braket_checkstyle_plugin.py (hunk 2 above)
Comment (krneta) [reply src/braket/flake8_plugins/braket_checkstyle_plugin.py:307]: Actually checking through the types would be better, but it would mean processing them twice or refactoring around how we deal with types. I'm not sure that it's worth it at this point.

**Review by ajberdy — APPROVED** [review]
